### PR TITLE
Add classes for IDs with slot index

### LINKED ISSUE #1947: bindings: add IDs slot index getter

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Add classes for IDs with slot index

## Links to any relevant issues

#1947 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Added tests, running wallet examples

### CODE REVIEW

File: bindings/nodejs/lib/types/block/id.ts [hunk 1]
```diff
@@ -2,6 +2,41 @@
...
+    }
+    inspect() {
+        return this.id;
+    }
+    toString() {
+        return this.id;
+    }
+    toJSON() {
+        return this.id;
+    }
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/block/id.ts]: Is there some way to have a class without named field? Haven't found a way

Comment (VmMad) [reply bindings/nodejs/lib/types/block/id.ts]: Did you mean to do something like this?
```typescript
class IdWithSlotIndex extends String {
    slotIndex(): SlotIndex {
        const numberString = super.toString().slice(66, 74);
        const chunks = [];
        for (
            let i = 0, charsLength = numberString.length;
            i < charsLength;
            i += 2
        ) {
            chunks.push(numberString.substring(i, i + 2));
        }
        const separated = chunks.map((n) => parseInt(n, 16));
        const buf = Uint8Array.from(separated).buffer;
        const view = new DataView(buf);
        return view.getUint32(0, true);
    }
}
```

if you do this you lose access to `.slotIndex()` if you do something like `new IdWithSlotIndex("0x..").toString()`

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: Maybe extend the String class and just add a method (tried this in devtools, seams to work)
```
class IdWithSlotIndex extends String {
    slotId() {
        this.length > 8 ? this.slice(-8) : this
    }
}
```
This way when you instantiate a "string" with `new IdWithSlotIndex("someBlockId")` you get the capabilities of the added function

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: Yes, thanks!

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: @msarcev but don't you see this as problem if it's available for all strings? Because most strings don't have a slot index encoded at the end

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: I don’t get the question 🤔 
It won’t be “available for all strings”, only strings that are instantiated with `new IdWithSlotIndex(…` , no ?

File: bindings/nodejs/lib/types/block/id.ts [hunk 2]
```diff
@@ -21,13 +43,18 @@ export type NftId = HexEncodedString;
 /**
  * A Block ID represented as hex-encoded string.
  */
-export type BlockId = HexEncodedString;
+export class BlockId extends IdWithSlotIndex {}
```
Comment (msarcev) [inline bindings/nodejs/lib/types/block/id.ts:46]: Here we could get away with `BlockId` still being a type
`type BlockId = IdWithSlotIndex` if that's better 🤔 
You won't be able to instantiate it as a class, but you should still be able to do:
```
const blockId: BlockId = new IdWithSlotIndex("12345678")
```
Again, not sure if it's better, just cleaner maybe.

File: bindings/nodejs/lib/types/block/id.ts (hunk 1 above)
Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: Ah, maybe you mean that you can put any string in there, even ones that are not "blockId/transactionId" 🤔 
That was also the case with your first solution though.
@VmMad suggested that we could add checks (maybe in the constructor) that assert if you passed something like a blockId (maybe by length). This would also mean handling errors if construction of that thing fails.
Wdyt ?

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: We could do something like:
```
export class IdWithSlotIndex extends String {
    private _isValid: boolean = true;

    constructor(id: string) {
        super(id)

        if (id.length < 64) {
            this._isValid = false;
        }
    }

    slotIndex(): SlotIndex {
        // ...
    }

    get isValid() {
        return this._isValid
    }
}
```
Then you can use it like:
const blockId = new IdWithSlotIndex("123123123")
```
if (blockId.isValid) {
    // use it
}
```

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: @Thoralf-M ☝🏻

File: bindings/nodejs/lib/types/block/id.ts (hunk 2 above)
Comment (VmMad) [reply bindings/nodejs/lib/types/block/id.ts:46]: Or maybe just rename `IdWithSlotIndex` to `BlockId`.

```tsx
  const slot = BlockId("0x...").slotIndex();
```

So it is also usable as type 

```typescript
  const blockId: BlockId = BlockId("0x...");
```

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts:46]: But then you can't call `.slotIndex()` on a blockid?

File: bindings/nodejs/lib/types/block/id.ts (hunk 1 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: > It won’t be “available for all strings”, only strings that are instantiated with new IdWithSlotIndex(… , no ?

Yes, sorry misread something

Sounds weird to me to have an `isValid()` method
But do I get it right that you prefer to use `new IdWithSlotIndex("0x...").slotIndex()` over `new BlockId("0x...").slotIndex()` if `BlockId` stays a type for a string instead of a class?

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts]: Maybe combine everything and instead of even having this `IdWithSlotIndex` class, make all "string" types that have a slot index encoded be a subclass of `String` with the added method explicitly.
```
class BlockId extends String {
    slotIndex() {
        // call to shared util
    }
}

class TransactionId extends String {
    slotIndex() {
        // call to shared util
    }
}
```
This way we can construct with `new BlockId("123123")` fine

File: bindings/nodejs/lib/types/block/id.ts (hunk 2 above)
Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts:46]: @Thoralf-M In which case you mean?
For the example i suggested up here you can do.
```
const blockId: BlockId = new IdWithSlotIndex("12345678")
blockId.slotIndex()
```
Cause the `BlockId` type is just an alias for `IdWithSlotIndex`

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts:46]: Now I understand, thanks
So do you prefer this, or https://github.com/iotaledger/iota-sdk/pull/1962#discussion_r1481491384? :smile:

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts:46]: At this point https://github.com/iotaledger/iota-sdk/pull/1962#discussion_r1481491384 is maybe the cleanest i think (no indirection by avoiding the extra type and it's clear that each "id thing" is a string class with extra util methods). @VmMad wdyt ?

Comment (VmMad) [reply bindings/nodejs/lib/types/block/id.ts:46]: @msarcev You know, at this point I think the solution @Thoralf-M did looks good to me. Because in case another method is added to the class `IdWithSlotIndex`, like `isValid()`, all the other classes that extend from it,  have it available, like `OutputId` or `SlotCommitmentId` which already extend `IdWithSlotIndex`. This way creating and referencing the util function multiple times in different classes.

TLDR; to me like it is done right now is fine, as they will be types correctly without needing an extra type
```tsx
export class BlockId extends IdWithSlotIndex {}
export class TransactionId extends IdWithSlotIndex {}
// ...
```

Comment (msarcev) [reply bindings/nodejs/lib/types/block/id.ts:46]: Agreed, we can go with this imo.

**Review by thibault-martinez — COMMENTED** [review]
- [x] iota-sdk/bindings/nodejs/lib/types/utils/bridge/utils.ts
- [x] iota-sdk/bindings/nodejs/lib/types/wallet/participation.ts
- [x] iota-sdk/bindings/nodejs/lib/types/wallet/signed-transaction-data.ts
- [x] iota-sdk/bindings/nodejs/lib/types/wallet/wallet.ts
- [x] iota-sdk/bindings/nodejs/lib/utils/utils.ts
- [x] iota-sdk/bindings/nodejs/lib/types/models/api/block-id-response.ts
- [x] iota-sdk/bindings/nodejs/lib/types/models/api/tips-response.ts
- [x] iota-sdk/bindings/nodejs/lib/types/models/block-metadata.ts
- [x] iota-sdk/bindings/nodejs/lib/types/client/prepared-transaction-data.ts
- [x] iota-sdk/bindings/nodejs/lib/types/models/api/plugins/indexer/output-response.ts

**Comment by Thoralf-M** [comment]
@thibault-martinez classes aren't allowed in map things `An index signature parameter type must be 'string', 'number', 'symbol', or a template literal type.ts(1268)`
Also wasn't sure about the other places because they're interfaces, but apparently that's fine, so changed these

File: bindings/nodejs/examples/client/05-get-address-balance.ts [hunk 3]
```diff
@@ -45,7 +51,9 @@ async function run() {
         });
 
         // Get outputs by their IDs
-        const addressOutputs = await client.getOutputs(outputIdsResponse.items);
+        const addressOutputs = await client.getOutputs(
+            outputIdsResponse.items.map((id) => new OutputId(id)),
```
Comment (thibault-martinez) [inline bindings/nodejs/examples/client/05-get-address-balance.ts:55]: No map needed anymore?

File: bindings/nodejs/examples/how_tos/client/get-outputs.ts [hunk 4]
```diff
@@ -33,7 +33,9 @@ async function run() {
         console.log('First output of query:');
         console.log('ID: ', outputIdsResponse.items[0]);
 
-        const outputs = await client.getOutputs(outputIdsResponse.items);
+        const outputs = await client.getOutputs(
+            outputIdsResponse.items.map((id) => new OutputId(id)),
```
Comment (thibault-martinez) [inline bindings/nodejs/examples/how_tos/client/get-outputs.ts:37]: No map needed anymore?

File: bindings/nodejs/examples/client/05-get-address-balance.ts (hunk 3 above)
Comment (Thoralf-M) [reply bindings/nodejs/examples/client/05-get-address-balance.ts:55]: Still needed because outputIdsResponse is an interface

Comment (thibault-martinez) [reply bindings/nodejs/examples/client/05-get-address-balance.ts:55]: Hmm ok, looks like a double wrap to me but if you say so

**Review by thibault-martinez — DISMISSED** [review]

File: bindings/nodejs/examples/client/05-get-address-balance.ts (hunk 3 above)
Comment (Thoralf-M) [reply bindings/nodejs/examples/client/05-get-address-balance.ts:55]: It's because the class is in an interface and then it's not actually this class object, this was also the reason why I initially didn't set this type there

**Review by Alex6323 — DISMISSED** [review]

File: bindings/nodejs/lib/types/block/id.ts [hunk 5]
```diff
@@ -2,6 +2,28 @@
...
 
 import { HexEncodedString } from '../utils';
+import { SlotIndex } from './slot';
+
+/**
+ * Base class for IDs with a hex encoded slot index at the end.
+ */
+export class IdWithSlotIndex extends String {
+    slotIndex(): SlotIndex {
+        const numberString = super.slice(66, 74);
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/block/id.ts]: use `length - 8` instead?

File: bindings/nodejs/lib/types/utils/bridge/utils.ts [hunk 6]
```diff
@@ -224,7 +224,7 @@ export interface __VerifyTransactionSemantic__ {
         inputs: InputSigningData[];
         protocolParameters: ProtocolParameters;
         unlocks?: Unlock[];
-        manaRewards?: { [outputId: OutputId]: NumericString };
+        manaRewards?: { [outputId: HexEncodedString]: NumericString };
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/utils/bridge/utils.ts:227]: we definitely can't use OutputId here? I know it can't be used in the dict but is there another way?

File: bindings/nodejs/lib/types/block/id.ts (hunk 5 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: No, there is the output index afterwards for output ids

File: bindings/nodejs/lib/types/utils/bridge/utils.ts (hunk 6 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/utils/bridge/utils.ts:227]: No idea what other way there could be

File: bindings/nodejs/lib/types/block/id.ts (hunk 5 above)
Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/id.ts]: sure but that should be called on the inner transaction ID not the output ID

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: Why? Then you would first need to get a transaction ID from the output ID

File: bindings/nodejs/lib/types/utils/bridge/utils.ts (hunk 6 above)
Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/utils/bridge/utils.ts:227]: there is a `Map` type in js, could probably use that instead: 
```js
manaRewards?: Map<OutputId, NumericString>;
```

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/utils/bridge/utils.ts:227]: I think there were some issues with the Map type, for example the serialization by default ends up in an empty object
And for things with a BTreeMap we also undid it here https://github.com/iotaledger/iota-sdk/pull/501

Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/utils/bridge/utils.ts:227]: :(

File: bindings/nodejs/lib/types/block/id.ts (hunk 5 above)
Comment (DaughterOfMars) [reply bindings/nodejs/lib/types/block/id.ts]: yeah, that's how we do it in rust

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/block/id.ts]: Changed

**Review by DaughterOfMars — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]
